pandas/core/dtypes/dtypes.py

```diff
@@ -494,18 +494,24 @@ def _hash_categories(self) -> int:
             cat_array = hash_tuples(cat_list)
         else:
             if categories.dtype == "O" and len({type(x) for x in categories}) != 1:
-                # TODO: hash_array doesn't handle mixed types. It casts
-                # everything to a str first, which means we treat
-                # {'1', '2'} the same as {'1', 2}
-                # find a better solution
-                hashed = hash((tuple(categories), ordered))
-                return hashed
-
-            if DatetimeTZDtype.is_dtype(categories.dtype):
-                # Avoid future warning.
-                categories = categories.view("datetime64[ns]")
-
-            cat_array = hash_array(np.asarray(categories), categorize=False)
+                if not ordered:
+                    # In case of mixed type, we don't want to directly convert values to
+                    # string or repr, as we may loose type information.Therefore we are
+                    # using {type}:{str(value)} as it's value representation for
+                    # hash calculation
+                    cat_array = hash_array(
+                        np.array([f"{str(type(cat))}:{str(cat)}" for cat in categories])
```

Comment:
I don't think this is the best idea. Since categories are unique, it's probably more appropriate to do a set comparison of the categories in `__eq__`

Reply:
Hi @mroeschke, Thanks for the suggestion! I have updated the logic to compare sets instead of hashes for object based categorical data.